src/Renderer/MenuRenderer.php

```diff
@@ -0,0 +1,147 @@
+<?php declare(strict_types=1);
+
+namespace Becklyn\Menu\Renderer;
+
+use Becklyn\Menu\Item\MenuItem;
+use Becklyn\Menu\Visitor\ItemVisitor;
+use Becklyn\Menu\Voter\VoterInterface;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Contracts\Service\ServiceSubscriberInterface;
+use Twig\Environment;
+
+class MenuRenderer implements ServiceSubscriberInterface
+{
+    /**
+     * @var ContainerInterface
+     */
+    private $locator;
+
+    /**
+     * @var ItemVisitor[]
+     */
+    private $visitors;
+
+
+    /**
+     * @var VoterInterface[]
+     */
+    private $voters;
+
+
+    /**
+     * @param ContainerInterface $locator
+     * @param ItemVisitor[]      $visitors
+     * @param VoterInterface[]   $voters
+     */
+    public function __construct (ContainerInterface $locator, iterable $visitors, iterable $voters)
+    {
+        $this->locator = $locator;
+        $this->visitors = $visitors;
+        $this->voters = $voters;
+    }
+
+
+    /**
+     * @param MenuItem $root
+     * @param array    $options
+     *
+     * @return string
+     */
+    public function render (MenuItem $root, array $options = []) : string
+    {
+        // don't modify the original
+        $root = clone $root;
+
+        // apply external visitors
+        // must be applied before voters, as they can generate new nodes
+        $this->applyVisitors($root);
```

Comment:
Does it sound like a good idea to you to pass in the `$options` to the visitors, as some decisions may be context-sensitive or even configurable via the options?

Reply:
I don't have a use case.
If you want something dependent on something else, then you can just pass them to the visitor directly.

Reply:
What about the currently active `Page`/`Website` in order to generate some items/entries?

Reply:
What do you want to do in a visitor that has to do with pages / websites?

I guess you might be confusing it with something like dynamic parameters (that we use in the RouteTreeBundle https://github.com/Becklyn/RouteTreeBundle#parameters), but they don't have anything with that here.
If you have these heavy deviations, you should just build separate trees. It's cheap enough.